### PR TITLE
LOOP-4665 Functional Algorithm

### PR DESCRIPTION
**Pull request by ps2**
https://tidepool.atlassian.net/browse/LOOP-4665

Adding a functional version of the Loop algorithm, to simplify testing, and simplify running of exploratory forecasts, like in the bolus calculator, without mucking with LoopDataManager state.

The new functional version is not used in any user facing code yet; it is simply run in a test to verify that it produces the same forecast as the existing LoopDataManager does for the same input data.

The input data is stored in a single file that contains all the diabetes data and Loop Algorithm settings needed to produce one forecast, and the expected output prediction is also stored in a fixture. Both the LoopDataManager and the functional LoopAlgorithm tests use the same fixtures.

* LoopAlgorithm.swift contains the new functional version of the prediction part of the Loop algorithm.
* New `LoopAlgorithmInput` type, which contains everything needed to run a prediction and dosing recommendation in the Loop Algorithm.
* Generic `GlucosePredictionAlgorithm` type added. Not currently used; just an illustration of how future alternate algorithms could conform to a common prototype.

Additional changes included (brought in to sync with DIY):

* `CarbValue` refactored to hold double, instead of HKQuantity, as we don't actually support any other units than grams.
* `StoredCarbEntry` refactored to allow some fields to be optional, which simplifies the fixtures that encode this type.  Also, the non-optional fields were moved to the beginning of the constructor.
* A few translations (not used in Tidepool Loop)
* Fix for time change crash. https://tidepool.atlassian.net/browse/LOOP-4723
* Fixes for favorite foods (not part of Tidepool Loop yet).

### CODE REVIEW

File: LoopKit/LoopAlgorithm/LoopAlgorithm.swift [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+            from: start.addingTimeInterval(-CarbMath.maximumAbsorptionTimeInterval).dateFlooredToTimeInterval(settings.delta),
+            to: nil)
+
+        // Future TODO: Calculate historic insulin effects for ICE at glucose sample timestamps. This will produce more accurate velocity samples.
+//        let effectDates = input.glucoseHistory.map { $0.startDate }
+//        let insulinEffectsAtGlucoseTimestamps = annotatedDoses.glucoseEffects(
+//            insulinModelProvider: insulinModelProvider,
+//            longestEffectDuration: input.insulinActivityDuration,
+//            insulinSensitivityTimeline: input.sensitivity,
+//            effectDates: effectDates)
```
Comment (nhamming) [inline LoopKit/LoopAlgorithm/LoopAlgorithm.swift]: remove?

File: LoopKit/LoopAlgorithm/LoopAlgorithm.swift [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+                retrospectiveCorrection: rcEffect,
+                momentum: momentumEffects,
+                insulinCounteraction: insulinCounteractionEffects
+            )
+        )
+    }
+
+//    // TODO: Generate recommendation from a prediction
+//    public static func generateRecommendation(prediction: LoopPrediction) throws -> LoopDosingRecommendation {
+//    }
```
Comment (nhamming) [inline LoopKit/LoopAlgorithm/LoopAlgorithm.swift]: remove?

**Review by nhamming — APPROVED** [review]
LGTM

File: LoopKit/LoopAlgorithm/LoopAlgorithm.swift (hunk 1 above)
Comment (ps2) [reply LoopKit/LoopAlgorithm/LoopAlgorithm.swift]: Thanks.  added https://tidepool.atlassian.net/browse/LOOP-4727 to keep track and will remove this commented code.

File: LoopKit/GlucoseKit/GlucoseMath.swift [hunk 3]
```diff
@@ -10,7 +10,7 @@ import Foundation
 import HealthKit
 
 public struct GlucoseMath {
-    public static let momentumDataInterval: TimeInterval = .minutes(15)
+    public static let momentumDataInterval: TimeInterval = .minutes(16)
```
Comment (Camji55) [inline LoopKit/GlucoseKit/GlucoseMath.swift]: Was this update intentional?

**Review by Camji55 — APPROVED** [review]

File: LoopKit/GlucoseKit/GlucoseMath.swift (hunk 3 above)
Comment (ps2) [reply LoopKit/GlucoseKit/GlucoseMath.swift:13]: Yes, but this probably deserves some more highlighting. This is fixing an issue where momentum is dropped due to cgm sensor timestamp jitter. I'll note this in a new ticket.

Comment (ps2) [reply LoopKit/GlucoseKit/GlucoseMath.swift]: Actually, going to revert this. This does change some behavior unintentionally.

**Pull request merged**